### PR TITLE
store/tikv: Support forwarding via follower when the region's leader is unreachable

### PR DESCRIPTION
**Pull request by MyonKeminta**
Signed-off-by: MyonKeminta <MyonKeminta@users.noreply.github.com>

### What problem does this PR solve?

Problem Summary:

Currently a region may be unavailable when the leader is network-isolated to the TiDB. In this case, we can try forwarding the request via a follower.

### What is changed and how it works?

What's Changed: Once the config `tikvclient.enable-redirection` is enabled, when it cannot connect to a TiKV node to access a region's leader, it will try asking a follower to forward the request to the leader.

### Related changes

- Need to cherry-pick to the release branch
  - Maybe 4.0
- Requires merging #23362 first

### Check List 

Tests (Not yet done)

- Unit test 
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Support forwarding region requests to the leader via a follower in case there's network isolation.

### CODE REVIEW

File: store/tikv/region_cache.go [hunk 1]
```diff
@@ -595,20 +631,35 @@ func (c *RegionCache) OnSendFail(bo *Backoffer, ctx *RPCContext, scheduleReload
...
+					logutil.BgLogger().Info("mark store's regions need be refill", zap.String("store", s.addr))
+					metrics.RegionCacheCounterWithInvalidateStoreRegionsOK.Inc()
+				}
 
-			// schedule a store addr resolve.
-			s.markNeedCheck(c.notifyCheckCh)
+				// schedule a store addr resolve.
+				s.markNeedCheck(c.notifyCheckCh)
+				// If the failing request is already being forwarded, do not start the loop
+				if EnableRedirection {
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:645]: Only requests sent to a region leader need redirection.

Comment (youjiali1995) [reply store/tikv/region_cache.go:645]: I prefer
```golang
		if err != nil {
			storeIdx, s := rs.accessStore(ctx.AccessMode, ctx.AccessIdx)
			leaderReq := ctx.Store.storeType == TiKV && rs.workTiKVIdx != ctx.AccessIdx

			//  Mark the store as failure if it's not a redirection request because we
			//  can't know the status of the proxy store by it.
			if ctx.ProxyStore == nil {
				// send fail but store is reachable, keep retry current peer for replica leader request.
				// but we still need switch peer for follower-read or learner-read(i.e. tiflash)
				if leaderReq {
					if s.requestLiveness(bo) == reachable {
						return
					} else if EnableRedirection {
						s.startHealthCheckLoopIfNeeded(c)
						startForwarding = true
					}
				}

				// invalidate regions in store.
				epoch := rs.storeEpochs[storeIdx]
				if atomic.CompareAndSwapUint32(&s.epoch, epoch, epoch+1) {
					logutil.BgLogger().Info("mark store's regions need be refill", zap.String("store", s.addr))
					metrics.RegionCacheCounterWithInvalidateStoreRegionsOK.Inc()
				}
				// schedule a store addr resolve.
				s.markNeedCheck(c.notifyCheckCh)
			}
		}
```

File: store/tikv/region_cache.go [hunk 2]
```diff
@@ -1406,6 +1480,37 @@ func (r *RegionStore) switchNextTiKVPeer(rr *Region, currentPeerIdx AccessIndex)
 	rr.compareAndSwapStore(r, newRegionStore)
 }
 
+func (r *RegionStore) switchNextProxyStore(rr *Region, currentProxyIdx AccessIndex) {
+	// The default value of `currentProxyIdx` is -1 which means proxy won't be used. Calling this function with -1 will
+	// increase the proxy store index to 0.
+
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:1486]: Move it to the upper of the method.

File: store/tikv/region_cache.go [hunk 3]
```diff
@@ -1406,6 +1480,37 @@ func (r *RegionStore) switchNextTiKVPeer(rr *Region, currentPeerIdx AccessIndex)
...
+	if r.proxyTiKVIdx != currentProxyIdx {
+		return
+	}
+	nextIdx := (currentProxyIdx + 1) % AccessIndex(r.accessStoreNum(TiKVOnly))
+	// skips the current workTiKVIdx
+	if nextIdx == r.workTiKVIdx {
+		nextIdx = (currentProxyIdx + 1) % AccessIndex(r.accessStoreNum(TiKVOnly))
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:1493]:
```suggestion
		nextIdx = (nextIdx + 1) % AccessIndex(r.accessStoreNum(TiKVOnly))
```

File: store/tikv/region_cache.go [hunk 4]
```diff
@@ -1643,6 +1753,55 @@ const (
...
+	}
+
+	// It may be already started by another thread.
+	if atomic.CompareAndSwapInt32(&s.needForwarding, 0, 1) {
+		go s.checkUntilHealth(c)
+	}
+}
+
+func (s *Store) checkUntilHealth(c *RegionCache) {
+	// TODO: Make the tick interval proper and configurable
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:1769]: I don't think it's necessary.

File: store/tikv/region_cache.go [hunk 5]
```diff
@@ -1643,6 +1753,55 @@ const (
...
+				storeMeta, err := c.PDClient().GetStore(ctx, s.storeID)
+				if err != nil {
+					logutil.BgLogger().Warn("failed to check unhealthy store's meta from pd", zap.Error(err))
+				}
+				if storeMeta == nil || storeMeta.State == metapb.StoreState_Tombstone {
+					// TODO: What to do here?
+					return
+				}
+			}
+
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:1792]:
```suggestion
			// The store is deleted due to metadata change. No need to check its health.
			if s.getResolveState() == deleted {
				return
			}
```

File: store/tikv/region_cache.go (hunk 1 above)
Comment (MyonKeminta) [reply store/tikv/region_cache.go]: It seems it should be `rs.workTiKVIdx == ctx.AccessIdx`

File: tidb-server/main.go [hunk 6]
```diff
@@ -589,6 +589,7 @@ func setGlobalVars() {
 
 	atomic.StoreUint64(&tikv.CommitMaxBackoff, uint64(parseDuration(cfg.TiKVClient.CommitTimeout).Seconds()*1000))
 	tikv.RegionCacheTTLSec = int64(cfg.TiKVClient.RegionCacheTTL)
+	tikv.EnableForwarding = cfg.EnableForwarding
```
Comment (youjiali1995) [inline tidb-server/main.go:592]: Is it a global config? That is to say, it's shared with PD.

File: store/tikv/region_request.go [hunk 7]
```diff
@@ -226,7 +227,8 @@ func (s *RegionRequestSender) SendReqCtx(
 	if span := opentracing.SpanFromContext(bo.ctx); span != nil && span.Tracer() != nil {
 		span1 := span.Tracer().StartSpan("regionRequest.SendReqCtx", opentracing.ChildOf(span.Context()))
 		defer span1.Finish()
-		bo = bo.Clone()
+		// TODO: Make sure trace works without cloning the backoffer.
```
Comment (youjiali1995) [inline store/tikv/region_request.go:230]: I prefer this way, otherwise, we don't know who is responsible for the TODO.
```suggestion
		// TODO(MyonKeminta): Make sure trace works without cloning the backoffer.
```

File: tidb-server/main.go (hunk 6 above)
Comment (MyonKeminta) [reply tidb-server/main.go:592]: Yes. It's introduced in https://github.com/pingcap/tidb/pull/23025 .

File: store/tikv/region_request.go (hunk 7 above)
Comment (MyonKeminta) [reply store/tikv/region_request.go:230]: Sounds good

File: store/tikv/region_cache.go [hunk 8]
```diff
@@ -365,6 +379,12 @@ func (c *RegionCache) GetTiKVRPCContext(bo *Backoffer, id RegionVerID, replicaRe
 		return nil, nil
 	}
 
+	if cachedRegion.needReload() {
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:382]: Why add it?

Comment (MyonKeminta) [reply store/tikv/region_cache.go:382]: In function `RegionRequestSender.SendReqCtx` there's a retry loop, in which it gets a RPCContext, send a request, and retry if sending failed. If it has tried all peer, it will mark the region as need reload, but reloading won't happen in this loop. So I added it here, and it will cause a fake `EpochNotMatch` error to the invoker of `SendReqCtx`, and the invoker will usually reload the region.

Comment (youjiali1995) [reply store/tikv/region_cache.go:382]: 🧐 I remembered there's someplace the error is escaped from the loop. Let me check it.

Comment (MyonKeminta) [reply store/tikv/region_cache.go:382]: When  getRPCContext returns nil, it will cause a empty EpochNotMatch error and excape from the loop..

File: store/tikv/region_request.go [hunk 9]
```diff
@@ -286,6 +288,7 @@ func (s *RegionRequestSender) SendReqCtx(
 
 			// TODO: Change the returned error to something like "region missing in cache",
 			// and handle this error like EpochNotMatch, which means to re-split the request and retry.
+			logutil.Logger(bo.ctx).Info("throwing pseudo region error due to region not founc in cache", zap.Stringer("region", &regionID))
```
Comment (youjiali1995) [inline store/tikv/region_request.go:291]: Can it print a lot? I don't think it's necessary.

File: store/tikv/region_request.go [hunk 10]
```diff
@@ -568,8 +592,22 @@ func (s *RegionRequestSender) NeedReloadRegion(ctx *RPCContext) (need bool) {
...
+	if ctx.AccessMode == TiKVOnly && len(s.failStoreIDs)+len(s.failProxyStoreIDs) >= ctx.TiKVNum {
+		need = true
+	} else if ctx.AccessMode == TiFlashOnly && len(s.failStoreIDs) >= len(ctx.Meta.Peers)-ctx.TiKVNum {
+		need = true
+	} else if len(s.failStoreIDs)+len(s.failProxyStoreIDs) >= len(ctx.Meta.Peers) {
+		need = true
+	}
+
 	if need {
 		s.failStoreIDs = nil
```
Comment (youjiali1995) [inline store/tikv/region_request.go:612]:
```suggestion
		s.failStoreIDs = nil
		s.failProxyStoreIDs = nil
```

File: store/tikv/region_request.go [hunk 11]
```diff
@@ -568,8 +592,22 @@ func (s *RegionRequestSender) NeedReloadRegion(ctx *RPCContext) (need bool) {
 	if s.failStoreIDs == nil {
 		s.failStoreIDs = make(map[uint64]struct{})
 	}
+	if s.failProxyStoreIDs == nil {
+		s.failProxyStoreIDs = make(map[uint64]struct{})
+	}
 	s.failStoreIDs[ctx.Store.storeID] = struct{}{}
```
Comment (youjiali1995) [inline store/tikv/region_request.go:598]: Why add it unconditionally? Why not
```golang
	if ctx.ProxyStore == nil {
		s.failStoreIDs[ctx.Store.storeID] = struct{}{}
	} else {
		s.failProxyStoreIDs[ctx.ProxyStore.storeID] = struct{}{}
	}
```

File: store/tikv/region_cache.go [hunk 12]
```diff
@@ -1642,18 +1795,84 @@ const (
...
+	}
+
+	// It may be already started by another thread.
+	if atomic.CompareAndSwapInt32(&s.needForwarding, 0, 1) {
+		go s.checkUntilHealth(c)
+	}
+}
+
+func (s *Store) checkUntilHealth(c *RegionCache) {
+	ticker := time.NewTicker(time.Millisecond * 500)
```
Comment (tiancaiamao) [inline store/tikv/region_cache.go:1813]: Is it too frequent, what's the old value?

File: store/tikv/region_cache.go [hunk 13]
```diff
@@ -1671,42 +1890,99 @@ func (s *Store) requestLiveness(bo *Backoffer) (l livenessState) {
...
 		} else {
 			metrics.StatusCountWithError.Inc()
 		}
-		metrics.TiKVStatusDuration.WithLabelValues(saddr).Observe(time.Since(start).Seconds())
+		metrics.TiKVStatusDuration.WithLabelValues(addr).Observe(time.Since(start).Seconds())
 	}()
 	ctx, cancel := context.WithTimeout(context.Background(), timeout)
 	defer cancel()
-	url := fmt.Sprintf("%s://%s/status", config.InternalHTTPSchema(), saddr)
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
```
Comment (tiancaiamao) [inline store/tikv/region_cache.go:1687]: You can make the change of liveness check a separate PR

File: store/tikv/region_cache.go [hunk 14]
```diff
@@ -1671,42 +1890,99 @@ func (s *Store) requestLiveness(bo *Backoffer) (l livenessState) {
...
 	}
-	resp, err := config.InternalHTTPClient().Do(req)
+	defer func() {
+		err := conn.Close()
+		if err != nil {
+			logutil.BgLogger().Info("[health check] failed to close the grpc connection for health check", zap.String("store", addr), zap.Error(err))
+		}
+	}()
+
+	req := &healthpb.HealthCheckRequest{}
```
Comment (tiancaiamao) [inline store/tikv/region_cache.go:1920]: This one introduced a protocol change, what happens when a new versioned TiDB works with an old versioned TiKV?

File: store/tikv/region_cache_test.go [hunk 15]
```diff
@@ -519,6 +520,58 @@ func (s *testRegionCacheSuite) TestSendFailInvalidateRegionsInSameStore(c *C) {
 	c.Assert(err, IsNil)
 }
 
+func (s *testRegionCacheSuite) TestSendFailEnableForwarding(c *C) {
+	EnableForwarding = true
```
Comment (tiancaiamao) [inline store/tikv/region_cache_test.go:524]: Test cases run parallelly, will this affect other tests?

File: store/tikv/region_request.go (hunk 9 above)
Comment (tiancaiamao) [inline store/tikv/region_request.go:291]: s/founc/found

File: store/tikv/region_request.go [hunk 16]
```diff
@@ -476,6 +489,16 @@ func (s *RegionRequestSender) sendReqToRegion(bo *Backoffer, rpcCtx *RPCContext,
...
 	}
 
+	if rpcCtx.ProxyStore != nil {
+		fromStore := strconv.FormatUint(rpcCtx.ProxyStore.storeID, 10)
+		toStore := strconv.FormatUint(rpcCtx.Store.storeID, 10)
+		result := "ok"
+		if err != nil {
+			result = "fail"
+		}
+		metrics.TiKVForwardRequestCounter.WithLabelValues(fromStore, toStore, result).Inc()
```
Comment (tiancaiamao) [inline store/tikv/region_request.go:499]: grafana json file is not changed?

File: store/tikv/region_cache.go [hunk 17]
```diff
@@ -1642,18 +1795,85 @@ const (
...
+				if err != nil {
+					logutil.BgLogger().Warn("[health check] failed to check unhealthy store's meta from pd", zap.Error(err))
+				}
+				if storeMeta == nil || storeMeta.State == metapb.StoreState_Tombstone {
+					// TODO(MyonKeminta): Should the region meta on this store be reloaded?
+					return
+				}
+			}
+			// The store is deleted due to metadata change. No need to check its health.
+			if s.getResolveState() == deleted {
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:1837]: It's not enough. When a store is `needForwarding`, it's seldom `reResolve`d.

Comment (youjiali1995) [reply store/tikv/region_cache.go:1837]: Ah, there is a timer to recheck all stores at the interval of 60s by default.

Comment (MyonKeminta) [reply store/tikv/region_cache.go:1837]: What should we do when it's not resolved yet? Should we skip until it's resolved or deleted?

File: store/tikv/region_request.go (hunk 11 above)
Comment (youjiali1995) [reply store/tikv/region_request.go:598]: I see. The `needForwarding` store is never chosen to send requests.

File: store/tikv/region_cache.go (hunk 17 above)
Comment (youjiali1995) [reply store/tikv/region_cache.go:1837]: Call `reResvole` in the loop.

File: store/tikv/region_cache.go [hunk 18]
```diff
@@ -1642,18 +1795,85 @@ const (
...
+		case <-ticker.C:
+			if time.Since(lastCheckPDTime) > time.Second*30 {
+				lastCheckPDTime = time.Now()
+
+				storeMeta, err := c.PDClient().GetStore(ctx, s.storeID)
+				if err != nil {
+					logutil.BgLogger().Warn("[health check] failed to check unhealthy store's meta from pd", zap.Error(err))
+				}
+				if storeMeta == nil || storeMeta.State == metapb.StoreState_Tombstone {
+					// TODO(MyonKeminta): Should the region meta on this store be reloaded?
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply store/tikv/region_cache.go:1832]: I'm not sure if it's necessary. But it seems it doesn't matter if we do it anyway

File: store/tikv/region_cache.go (hunk 12 above)
Comment (MyonKeminta) [reply store/tikv/region_cache.go:1813]: It's new code and there's no old value. I think I'll try to increase it to 1s. Is it ok?

File: store/tikv/region_request.go (hunk 9 above)
Comment (MyonKeminta) [reply store/tikv/region_request.go:291]: Maybe debug level is ok

File: store/tikv/region_cache.go (hunk 14 above)
Comment (MyonKeminta) [reply store/tikv/region_cache.go:1920]: An unhealthy store will never become healthy. But is this situation common where new TiDB runs together with old version TiKV?

File: store/tikv/region_cache_test.go (hunk 15 above)
Comment (MyonKeminta) [reply store/tikv/region_cache_test.go:524]: Will it run parallelly with others even testRegionCacheSuite inherits OneByOneSuite?

File: store/tikv/region_cache.go (hunk 17 above)
Comment (MyonKeminta) [reply store/tikv/region_cache.go:1837]: I'm afraid this won't work as we expected. After resolving it creates a new `Store` object and discards the old one. I think in this case the loop will exit since the `Store`'s state becomes deleted, and another access to this store will trigger the check loop on the new `Store` object.

File: store/tikv/region_request_test.go [hunk 19]
```diff
@@ -569,3 +572,129 @@ func (s *testRegionRequestToSingleStoreSuite) TestOnMaxTimestampNotSyncedError(c
...
+	}}
+
+	loc, err := s.regionRequestSender.regionCache.LocateKey(bo, []byte("k"))
+	c.Assert(err, IsNil)
+	c.Assert(loc.Region.GetID(), Equals, s.regionID)
+	req := tikvrpc.NewRequest(tikvrpc.CmdRawPut, &kvrpcpb.RawPutRequest{
+		Key:   []byte("k"),
+		Value: []byte("v1"),
+	})
+	resp, ctx, err := s.regionRequestSender.SendReqCtx(bo, req, loc.Region, time.Second, kv.TiKV)
```
Comment (youjiali1995) [inline store/tikv/region_request_test.go:615]: Trigger request forwarding because `mocktikv` doesn't implement the health service? I prefer to use the failpoint here.
```golang
	c.Assert(failpoint.Enable("github.com/pingcap/tidb/store/tikv/mockRequestLiveness", "return(false)"), IsNil)
```

File: store/tikv/region_request_test.go [hunk 20]
```diff
@@ -569,3 +572,129 @@ func (s *testRegionRequestToSingleStoreSuite) TestOnMaxTimestampNotSyncedError(c
...
+	c.Assert(failpoint.Disable("github.com/pingcap/tidb/store/tikv/mockRequestLiveness"), IsNil)
+
+	req = tikvrpc.NewRequest(tikvrpc.CmdRawGet, &kvrpcpb.RawGetRequest{Key: []byte("k")})
+	resp, ctx, err = s.regionRequestSender.SendReqCtx(bo, req, loc.Region, time.Second, kv.TiKV)
+	c.Assert(err, IsNil)
+	regionErr, err = resp.GetRegionError()
+	c.Assert(err, IsNil)
+	c.Assert(regionErr, IsNil)
+	c.Assert(resp.Resp.(*kvrpcpb.RawGetResponse).Value, BytesEquals, []byte("v1"))
+
```
Comment (youjiali1995) [inline store/tikv/region_request_test.go:647]: Should check `RPCContext` here.

File: store/tikv/region_request_test.go [hunk 21]
```diff
@@ -569,3 +572,129 @@ func (s *testRegionRequestToSingleStoreSuite) TestOnMaxTimestampNotSyncedError(c
...
+	for {
+		if atomic.LoadInt32(&leaderStore.needForwarding) == 0 {
+			break
+		}
+		if time.Since(start) > 3*time.Second {
+			c.Fatal("store didn't recover to normal in time")
+		}
+	}
+	c.Assert(failpoint.Disable("github.com/pingcap/tidb/store/tikv/mockRequestLiveness"), IsNil)
+
```
Comment (youjiali1995) [inline store/tikv/region_request_test.go:639]: Sleep 2s because the interval of the health check is 1s.

File: store/tikv/region_cache_test.go (hunk 15 above)
Comment (tiancaiamao) [reply store/tikv/region_cache_test.go:524]: By default, each suite runs parallelly. 
After declaring OneByOneSuite, those suites in the OneByOneSuite list run one by one.
Not all the test suites are declared as OneByOneSuite.
So I'm not sure will this change make the CI non-stable.

File: store/tikv/region_cache.go (hunk 14 above)
Comment (tiancaiamao) [reply store/tikv/region_cache.go:1936]: It may affect the rolling update.
Seems not a big problem.

File: store/tikv/region_cache.go (hunk 12 above)
Comment (tiancaiamao) [reply store/tikv/region_cache.go]: OKay, and notice the log printing, avoiding too many annoying logs that quickly exhaust the disk.

File: store/tikv/region_request_test.go [hunk 22]
```diff
@@ -569,3 +572,129 @@ func (s *testRegionRequestToSingleStoreSuite) TestOnMaxTimestampNotSyncedError(c
...
+	})
+	resp, ctx, err := s.regionRequestSender.SendReqCtx(bo, req, loc.Region, time.Second, kv.TiKV)
+	c.Assert(err, IsNil)
+	regionErr, err := resp.GetRegionError()
+	c.Assert(err, IsNil)
+	c.Assert(regionErr, IsNil)
+	c.Assert(resp.Resp.(*kvrpcpb.RawPutResponse).Error, Equals, "")
+	c.Assert(ctx.Addr, Equals, leaderAddr)
+	c.Assert(ctx.ProxyStore, NotNil)
+	c.Assert(ctx.ProxyAddr, Not(Equals), leaderAddr)
```
Comment (youjiali1995) [inline store/tikv/region_request_test.go:623]:
```suggestion
	c.Assert(ctx.ProxyAddr, Not(Equals), leaderAddr)
	c.Assert(ctx.ProxyAccessIdx, Not(Equals), ctx.AccessIdx)
```

File: store/tikv/region_request_test.go [hunk 23]
```diff
@@ -569,3 +572,129 @@ func (s *testRegionRequestToSingleStoreSuite) TestOnMaxTimestampNotSyncedError(c
...
+	})
+	resp, ctx, err = s.regionRequestSender.SendReqCtx(bo, req, loc.Region, time.Second, kv.TiKV)
+	c.Assert(err, IsNil)
+	regionErr, err = resp.GetRegionError()
+	c.Assert(err, IsNil)
+	// After several retries, the region will be marked as needReload.
+	// Then SendReqCtx will throw a pseudo EpochNotMatch to tell the caller to reload the region.
+	c.Assert(regionErr.EpochNotMatch, NotNil)
+	c.Assert(ctx, IsNil)
+
```
Comment (youjiali1995) [inline store/tikv/region_request_test.go:681]:
```suggestion
	c.Assert(len(s.regionRequestSender.failStoreIDs), Equals, 0)
	c.Assert(len(s.regionRequestSender.failProxyStoreIDs), Equals, 0)
	region := s.regionRequestSender.regionCache.getCachedRegionWithRLock(loc.Region)
	c.Assert(region, NotNil)
	c.Assert(region.checkNeedReload(), IsTrue)

```

File: store/tikv/region_request_test.go (hunk 19 above)
Comment (MyonKeminta) [reply store/tikv/region_request_test.go:615]: It's hard to let the failpoint know what address to refuse, and whether forwarded reqeusts should success, to do this we need to pass there information with a string, and parse it in the failpoint. So I think the `fnClient` (which is already provided) is more easy to use.

File: store/tikv/region_request_test.go (hunk 21 above)
Comment (MyonKeminta) [reply store/tikv/region_request_test.go:639]: There's a retry loop that will check it for up to 3s above.

File: store/tikv/region_request_test.go (hunk 19 above)
Comment (MyonKeminta) [reply store/tikv/region_request_test.go:613]: And `mockRequestLiveness` only injects the `requestLiveness` function, which won't be executed if no RPC error occurs.

Comment (youjiali1995) [reply store/tikv/region_request_test.go:613]: My point is when `fnClient` reports error, the `requestLiveness` fails because the `mocktikv` doesn't implement health service now. I prefer to use the fail point to make `requestLiveness` fail.

File: store/tikv/region_cache_test.go [hunk 24]
```diff
@@ -519,6 +520,59 @@ func (s *testRegionCacheSuite) TestSendFailInvalidateRegionsInSameStore(c *C) {
...
+	c.Assert(err, IsNil)
+	c.Assert(loc2.Region.id, Equals, region2)
+	ctx, err = s.cache.GetTiKVRPCContext(s.bo, loc2.Region, kv.ReplicaReadLeader, 0)
+	c.Assert(err, IsNil)
+	c.Assert(ctx, NotNil)
+	c.Assert(ctx.ProxyStore, NotNil)
+	c.Assert(ctx.ProxyStore.storeID, Equals, s.store2)
+
+	// Recover the store
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/store/tikv/mockRequestLiveness", "return(true)"), IsNil)
```
Comment (tiancaiamao) [inline store/tikv/region_cache_test.go:560]: Is this still used?

File: store/tikv/region_cache.go (hunk 14 above)
Comment (youjiali1995) [reply store/tikv/region_cache.go:1934]: Nope. In the normal process of upgrading, PD's version >= TiKV's version >= TiDB's version.

File: store/tikv/region_cache_test.go (hunk 24 above)
Comment (MyonKeminta) [reply store/tikv/region_cache_test.go]: Sorry I forgot  to remove it

File: store/tikv/region_request_test.go (hunk 19 above)
Comment (youjiali1995) [reply store/tikv/region_request_test.go:613]: Never mind. The fail point is removed. 🤣

File: store/tikv/region_cache.go [hunk 25]
```diff
@@ -365,6 +393,13 @@ func (c *RegionCache) GetTiKVRPCContext(bo *Backoffer, id RegionVerID, replicaRe
 		return nil, nil
 	}
 
+	if cachedRegion.checkNeedReload() {
+		logutil.BgLogger().Info("return nil on needReload")
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:397]: Is it necessary?

**Review by youjiali1995 — APPROVED** [review]

**Review by youjiali1995 — APPROVED** [review]

**Review by youjiali1995 — APPROVED** [review]
